Handling of PHP's automatic array key conversion to int

peterfabian opened the pull request:
Fixes #19619 

Testing instructions:

1. Create a product with id <product_id>, set two other products as upsells (e.g. with ids 77 and 48).
2. Creat a postmeta with an array of integers represented as strings, such as this:
```
array(
	0 => "77",
	1 => "48"
)
```
e.g. by running
```
UPDATE 
	`wp_postmeta` 
SET 
	`meta_value` = 'a:2:{i:0;s:2:"77";i:1;s:2:"48";}' 
WHERE 
	`wp_postmeta`.`post_id` = <product_id> 
	AND `wp_postmeta`.`meta_key` = '_purchase_note';
```

4. Create a multiselect for `_purchase_note`, e.g.:
```
add_action( 'woocommerce_product_options_general_product_data', 'woo_add_custom_general_fields' );
function woo_add_custom_general_fields() {
	global $post;

	// <option value="#post_id#">#post_name#</option>
	$options = array();
	$values = get_post_meta( $post->ID, '_upsell_ids', true );
	foreach( $values as $key => $value ) {
		$prod = WC()->product_factory->get_product( $value );
		$options[$value] = $prod->get_name();
	}

	echo '<div class="options_group">';

	woocommerce_wp_select( array(
		'id'      => '_upsell_ids',
		'label'   => __( '_upsell_ids', 'woocommerce' ),
		'options' =>  $options,
		'custom_attributes' => array('multiple' => ''),
	) );

	echo '</div>';
}
```
5. Observe the original code does not have any selected upsell product names.
6. Apply code change and observe the upsell products are selected.

Updated testing instructions for refactored methods:
- WC > Coupons > Edit coupon > Usage restrictions > Product categories & Exclude categories -- both selects should show picked options correctly. 
- WC > Settings > General > Country/State -- should show picked option correctly
- WC > Settings > Shipping > Zone > Edit > Zone regions -- should show picked option correctly
- Products > Edit product > Attributes -- should show picked attributes correctly
- WC > Reports > Coupons by date > Filter by coupon functionality works correctly

peterfabian committed:
Handling of PHP's automatic array key conversion to int

peterfabian committed:
Improved the comment

ragulka commented on includes/admin/wc-meta-box-functions.php:
```diff
@@ -227,7 +227,10 @@ function woocommerce_wp_select( $field ) {
+				// The $key from $field['options'] could have been converted to integer automatically by PHP,
+				// thus the comparison needs to check both $key and (string) $key in case $field['value'] is an array
+				// with string values.
+				echo '<option value="' . esc_attr( $key ) . '" ' . selected( $field['value'] === $key || ( is_array( $field['value'] ) && ( in_array( $key, $field['value'], true ) || in_array( (string) $key, $field['value'], true ) ) ), true, false ) . '>' . esc_html( $value ) . '</option>';
```
Would it be possible to make sure that the comparison also works for regular select fields? See my comment at https://github.com/woocommerce/woocommerce/issues/19619#issuecomment-381055290 for more details.

peterfabian replied:
Thx, that's a good point, I missed that since I focused on the multiselect problem.

mikejolley replied:
@peterfabian Cast both? As we said, `$key` will be a string when posted. So if we cast the $key and the compared value we only need one set of logic.

peterfabian replied:
Well, `meta_value` is a text, so I thought it's safe. But it's true that if value is passed manually instead of fetched from the database, then it would still fail.

peterfabian committed:
Fix strict comparison for normal select, too.

peterfabian commented:
I think that would break if `field['value']` is an array of integers, e.g. `_upsell_ids` can be `a:2:{i:0;i:38;i:1;i:197;}`, then strict comparison would fail as string $key would not be found in array `$field['value']`. 
Or do you think we should also convert all values in `$field['value']` to strings and then compare `$key` converted to string?

peterfabian commented:
So... should we just abandon strict checking altogether and go for `==` and `in_arrray(_, _, false)`?

mikejolley commented:
> Or do you think we should also convert all values in $field['value'] to strings and then compare $key converted to string?

Thats my suggestion. Can you think of a case that would fail?

peterfabian committed:
Converted all input values to selected to strings

Extracted conversions and compare logic to a separate function to improve readability.

peterfabian commented:
Updated, please check.

peterfabian committed:
Added wc prefix to function name.

peterfabian committed:
Specified input parameters more precisely.

mikejolley commented on includes/admin/wc-meta-box-functions.php:
```diff
@@ -181,6 +181,26 @@ function woocommerce_wp_checkbox( $field ) {
+ * @param string|int|array $options Options to go through when looking for value.
+ * @return bool
+ */
+function wc_is_value_in_options( $value, $options ) {
```
👍 

Since this is specialised in nature how about we replace `selected` itself. I.e. create `wc_selected` which applies your new logic and returns a string - `selected` if true, empty if false. That takes care of markup too and should be cleaner overall.

A unit test for that helper function would also be useful.

claudiulodro replied:
+1 for unit test for this function Unit tests are a good way of documenting the correct behavior so we don't accidentally break this in the future. 

Since this is a bug fix it can be merged before or after the 3.4 beta is tagged, so you have time.

mikejolley reviewed (commented):
Added an idea

peterfabian committed:
Added wc_selected function.

- Added unit test for the wc_selected function.

peterfabian committed:
Changed usage of selected with in_array to wc_selected.

claudiosanches commented on includes/admin/wc-meta-box-functions.php:
```diff
@@ -181,6 +181,33 @@ function woocommerce_wp_checkbox( $field ) {
+	}
+
+	if ( $is_value_in_options ) {
+		return " selected='selected'";
```
No need for double quotes here, we can just use `' selected="selected"';`. WP returns in single quotes because of https://developer.wordpress.org/reference/functions/__checked_selected_helper/ but we don't need it here.

mikejolley replied:
I think it's find to just return `selected` - thats valid HTML5

claudiosanches commented on includes/admin/wc-meta-box-functions.php:
```diff
@@ -181,6 +181,33 @@ function woocommerce_wp_checkbox( $field ) {
+ * @param string|int|array $options Options to go through when looking for value.
+ * @return string
+ */
+function wc_selected( $value, $options ) {
```
You can resume this functions to:

```php
function wc_selected( $value, $options ) {
	if ( is_array( $options ) ) {
		$options = array_map( 'strval', $options );
		return selected( in_array( (string) $value, $options, true ), true, false );
	}

	return selected( $value, $options, false );
}
```

So you can ignored the next comment about `selected='selected'`

claudiosanches reviewed (dismissed):
Also this function is too generic to be on admin meta box functions, I think we can put as core function.

peterfabian committed:
Used the selected function to take care of markup and simple case.

peterfabian commented:
@claudiosanches happy to move it as it also felt quite generic for that module, any candidate modules you think this would fit in? wc-core-functions.php?

mikejolley commented:
Looks good now 👍 

I'm happy to leave it here @claudiosanches. It's only used in meta boxes. If we ever come across a use case for it in other places, such as frontend, we can move it to core functions.

🚢

peterfabian committed:
Moved wc_selected to core functions.

peterfabian commented:
@mikejolley Moved and retested quickly, seems to work fine.
